### PR TITLE
fix: auto-recover zero-throughput watcher wedges

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary

- add a dedicated 60-second `com.brainlayer.throughput-watchdog` LaunchAgent
- compare the indexed `realtime_watcher` high-water rowid with recent registry-tracked JSONL bytes that remain ahead of confirmed offsets
- require three consecutive zero-throughput ticks with positive pending evidence before recovery
- alert before a checked `launchctl kickstart -k`, verify the label afterward, and apply a ten-minute recovery cooldown
- install the standalone bounded probe through the packaged launchd installer

## Why

The third production incident left `com.brainlayer.watch` alive but wedged in an uninterruptible SQLite wait. Process liveness and launchd KeepAlive therefore remained green while ingestion stopped. Recovery must use throughput plus pending-input evidence.

## Verification

- focused launchd/install/watchdog/tier-0 suite: 75 passed
- Ruff check and format: passed
- `bash -n`, ShellCheck, plist lint, and `git diff --check`: passed
- live dry-run against the 687k-row production DB and 13.7k-entry offset registry: 0.44-1.64 seconds
- live pre-deploy evidence: watcher high-water rowid 985223 remained flat while 5 tracked files held about 4.5 MB beyond confirmed offsets

## Deployment gate

This is deploy-bearing. After merge, install and load the LaunchAgent, then perform the authorized forced-wedge drill: stop the live watcher without exiting it, create positive pending-input evidence, prove the watchdog replaces the PID automatically, and prove watcher chunks resume. LayerSpec owns merge timing.

---

> [!NOTE]
> **Medium Risk**
> Production ingestion recovery now depends on automatic `launchctl kickstart -k` and filesystem/DB heuristics; mis-stalls could restart a healthy watcher, though thresholds, cooldown, and liveness checks mitigate false positives.
> 
> **Overview**
> Adds a **throughput watchdog** for when `com.brainlayer.watch` stays alive but stops ingesting—complementing process-level checks like tier0-watchdog.
> 
> A new **`throughput-watchdog.py`** runs on a **60s** LaunchAgent schedule. Each tick compares SQLite **`realtime_watcher`** chunk high-water and **`watcher_liveness_events`** progress against recent JSONL under the usual session roots (with bounded `find` and offset-registry logic aligned to watcher Cursor paths). If pending bytes remain but throughput is flat for **three** consecutive checks, it alerts, persists recovery state, **`launchctl kickstart -k`** the watch job (with post-kickstart running verification), and enforces a **10-minute** cooldown. A file lock avoids overlapping runs; probe failures fail closed rather than treating a bad scan as idle.
> 
> **`install.sh`** gains `throughput-watchdog` install/remove, includes it in **`all`**, and ships **`com.brainlayer.throughput-watchdog.plist`** via the packaged wheel. Tests cover stall/recovery edge cases, installer rendering, and launchd hygiene.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 44d41786958ce53a5b9d4077b91e55f067be34fb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add throughput watchdog to auto-recover zero-throughput realtime watcher stalls
> - Adds [throughput-watchdog.py](https://github.com/EtanHey/brainlayer/pull/600/files#diff-ce23882a33c4aca82486682d2aded2534de2839e67bbb05ef8b09c0c0e7063e6), a new Python CLI that detects wedged realtime watcher processes by comparing SQLite DB highwater marks and on-disk JSONL source evidence across successive runs.
> - When the watcher appears stalled beyond a configurable threshold, the watchdog issues `launchctl kickstart` (with bootstrap if needed) and polls until the job reaches `running` state, with cooldown and dry-run support.
> - Adds [com.brainlayer.throughput-watchdog.plist](https://github.com/EtanHey/brainlayer/pull/600/files#diff-f7f0161ea3523ee4f664f0308cf63e4841ba7462b54a10fd9770fafdfbdd062d) as a LaunchAgent that runs the watchdog every 60 seconds under the BrainLayer env loader, logging to per-user BrainLayer logs.
> - Extends [install.sh](https://github.com/EtanHey/brainlayer/pull/600/files#diff-c2e9ec925e37d2fc11ea258b25bb7f65308a6755174d1b556a6c077d798bf756) with an `install_throughput_watchdog` function and wires it into the `all` and `remove` targets.
> - Risk: the watchdog uses a file lock to prevent concurrent runs, but a stale lock file could skip a recovery cycle.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 44d4178.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added an automated throughput watchdog that monitors realtime processing and detects stalled activity.
  * Automatically attempts recovery when processing remains stalled, with configurable notifications and logging.
  * Added installation and removal support for the new background service.
  * The watchdog runs automatically at regular intervals and supports status output in JSON format.

### CODE REVIEW

File: scripts/launchd/throughput-watchdog.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+                "-print0",
+            ],
+            capture_output=True,
+            timeout=config.max_scan_seconds,
+            check=False,
+        )
+    except subprocess.TimeoutExpired as exc:
+        raise RuntimeError(f"recent-source scan exceeded {config.max_scan_seconds:.1f}s") from exc
+    if completed.returncode != 0:
+        scan_errors += 1
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 8dbcdd86: any nonzero find result now raises with stderr instead of treating partial stdout as authoritative zero evidence. A regression forces find exit 1 and proves the probe fails visibly rather than resetting the stall counter.

File: scripts/launchd/throughput-watchdog.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+    kickstart_returncode, kickstart_output = _command_result(
+        command_runner,
+        ["launchctl", "kickstart", "-k", target],
+    )
+    if kickstart_returncode != 0:
+        raise RuntimeError(f"launchctl kickstart failed: {kickstart_output or kickstart_returncode}")
+    verify_returncode, verify_output = _command_result(command_runner, ["launchctl", "print", target])
+    if verify_returncode != 0:
+        raise RuntimeError(f"launchctl print after kickstart failed: {verify_output or verify_returncode}")
+    return f"kickstart:{config.watch_label}"
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 1bc7164b: after kickstart the watchdog polls launchctl print boundedly and accepts recovery only after state = running. A permanently exited fake job now produces recovery_failed and retains cooldown protection.

File: scripts/launchd/throughput-watchdog.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+        next_state["last_restart_epoch"] = checked_at
+        next_state["restart_count"] = int(state.get("restart_count", 0)) + 1
+        next_state.pop("last_recovery_error", None)
+    elif result.action == "recovery_failed":
+        next_state["last_recovery_error"] = result.error
+    if result.alert_error:
+        next_state["last_alert_error"] = result.alert_error
+    else:
+        next_state.pop("last_alert_error", None)
+    _atomic_write_json(config.state_path, next_state)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py:552]: Fixed in 1bc7164b: recovery_attempt, last_restart_epoch, evidence, and attempt count are atomically written and fsynced before kickstart. Tests verify ordering and prove the durable attempt survives a post-kickstart final state-write failure.

File: scripts/launchd/throughput-watchdog.py [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+    database = db_path.expanduser().resolve()
+    if not database.is_file():
+        raise RuntimeError(f"BrainLayer DB does not exist: {database}")
+    uri = f"{database.as_uri()}?mode=ro"
+    try:
+        with sqlite3.connect(uri, uri=True, timeout=2.0) as connection:
+            connection.execute("PRAGMA query_only = ON")
+            row = connection.execute(
+                "SELECT COALESCE(MAX(rowid), 0) FROM chunks WHERE source = 'realtime_watcher'"
+            ).fetchone()
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 1bc7164b: the progress token now advances on either the realtime_watcher chunk rowid or watcher_liveness_events rowid. A dedupe-only ingestion regression proves liveness progress resets the stall decision without a kickstart.

File: scripts/launchd/throughput-watchdog.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+            check=False,
+        )
+    except subprocess.TimeoutExpired as exc:
+        raise RuntimeError(f"recent-source scan exceeded {config.max_scan_seconds:.1f}s") from exc
+    if completed.returncode != 0:
+        detail = os.fsdecode(completed.stderr).strip()
+        raise RuntimeError(f"recent-source scan failed: {detail or completed.returncode}")
+    recent_paths = [Path(os.fsdecode(raw)) for raw in completed.stdout.split(b"\0") if raw]
+    if len(recent_paths) > config.max_source_files:
+        raise RuntimeError(f"recent-source scan exceeded {config.max_source_files} JSONL files")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 1bc7164b: find stdout/stderr are streamed through a selector, NUL paths are counted incrementally, and the process is terminated immediately at max_source_files + 1. Nonzero exit and timeout behavior remain fail-closed and covered.

File: scripts/launchd/throughput-watchdog.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+
+    for source_file in recent_paths:
+        if time.monotonic() - started > config.max_scan_seconds:
+            raise RuntimeError(f"source evidence exceeded {config.max_scan_seconds:.1f}s")
+        try:
+            source_stat = source_file.stat()
+        except OSError:
+            scan_errors += 1
+            continue
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 1bc7164b: any OSError while statting a recent candidate raises the probe error, so partial source evidence can never authorize recovery. Regression coverage forces PermissionError and verifies visible failure.

File: scripts/launchd/throughput-watchdog.py [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+        except OSError as exc:
+            raise RuntimeError(f"cannot inspect recent source file {source_file}: {exc}") from exc
+        if source_stat.st_mtime < cutoff:
+            continue
+        recent_files += 1
+        newest_mtime = source_stat.st_mtime if newest_mtime is None else max(newest_mtime, source_stat.st_mtime)
+        offset = _registry_offset(registry.get(str(source_file)), current_inode=source_stat.st_ino)
+        if offset is None:
+            untracked_recent_files += 1
+            continue
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py:323]: Fixed in a52e09fc. Recent non-empty JSONLs that match the watcher discovery surface now count as pending even before an offset entry exists, so a wedge before first discovery cannot look idle. The evidence scan also mirrors the narrower Cursor projects `agent-transcripts` rule, preventing unsupported project JSONLs from creating false recovery pressure. Covered by focused regressions.

File: scripts/launchd/com.brainlayer.throughput-watchdog.plist [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>Label</key>
+	<string>com.brainlayer.throughput-watchdog</string>
+	<key>ProgramArguments</key>
+	<array>
+		<string>__PYTHON_BIN__</string>
+		<string>__THROUGHPUT_WATCHDOG_SCRIPT__</string>
+		<string>--json</string>
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/com.brainlayer.throughput-watchdog.plist:12]: Fixed in a52e09fc. The LaunchAgent now invokes the installed script through `brainlayer-env-run.sh`, loads the canonical `BRAINLAYER_ENV_FILE`, and uses the `watch` service gate. This keeps `BRAINLAYER_DB`, the offset registry location derived from it, and system/watcher disable state aligned. Installer and launchd hygiene regressions cover the rendered job.

File: scripts/launchd/throughput-watchdog.py [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+    return result
+
+
+def _build_parser() -> argparse.ArgumentParser:
+    home = Path.home()
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument(
+        "--db", type=Path, default=Path(os.environ.get("BRAINLAYER_DB", home / ".local/share/brainlayer/brainlayer.db"))
+    )
+    parser.add_argument("--registry", type=Path, default=home / ".local/share/brainlayer/offsets.json")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 44d41786. `--registry` now defaults after argument parsing to `selected_db.parent / offsets.json`, so both `BRAINLAYER_DB` and explicit `--db` select the same progress/offset namespace unless a registry override is explicitly supplied. Regression covers explicit custom DB selection.

File: scripts/launchd/throughput-watchdog.py [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+            raise RuntimeError(f"cannot inspect recent source file {source_file}: {exc}") from exc
+        if source_stat.st_mtime < cutoff:
+            continue
+        recent_files += 1
+        newest_mtime = source_stat.st_mtime if newest_mtime is None else max(newest_mtime, source_stat.st_mtime)
+        offset = _registry_offset(registry.get(str(source_file)), current_inode=source_stat.st_ino)
+        if offset is None:
+            untracked_recent_files += 1
+            continue
+        if source_stat.st_size > offset:
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/throughput-watchdog.py]: Fixed in 44d41786. A recent tracked file whose size is below its stored offset is now pending rewind evidence; the full current size is counted for reread, and even a zero-byte rewind retains a pending-file signal. Regression covers the in-place truncation case.
